### PR TITLE
fix(setprops): allowed for setProps to be synced with nextTick intervals

### LINKED ISSUE #1419: Watch with immediate totally breaks setProps
### Version
1.0.0-beta.31

### Reproduction link
[https://github.com/aldarund/vue-test-utils-transition-bug/blob/master/components/watch.test.js](https://github.com/aldarund/vue-test-utils-transition-bug/blob/master/components/watch.test.js)
https://github.com/aldarund/vue-test-utils-transition-bug/blob/master/components/watchtest.vue

### Steps to reproduce
use watch with immediate as prop

### What is expected?
both test run succesfully

### What is actually happening?
setProps when there is watch with immediate in component dont do anything, it just dont set prop at all.

### PR DESCRIPTION
**Pull request by AtofStryker**
setProps in certain cases was being blown away by nextTick intervals. If the property is not up to
date, setProps will be called again to sync the changes.

fix #1419

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] It's submitted to the `dev` branch.
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing: https://github.com/vuejs/vue-test-utils/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: packages/test-utils/src/wrapper.js [hunk 1]
```diff
@@ -711,7 +711,15 @@ export default class Wrapper implements BaseWrapper {
 
       // $FlowIgnore : Problem with possibly null this.vm
       this.vm.$forceUpdate()
-      return nextTick()
+      return new Promise((resolve, reject) => {
+        nextTick().then(() => {
+          const isUpdated = Object.keys(data).some(key => {
+            // $FlowIgnore : Problem with possibly null this.vm
+            return this.vm[key] === data[key]
```
Comment (AtofStryker) [inline packages/test-utils/src/wrapper.js]: not sure if we need to compare by value here, but figured this is handled above? Might also be a if the same object property is set, but another is not.

Comment (lmiller1990) [reply packages/test-utils/src/wrapper.js]: This is comparing by value right?

**Review by lmiller1990 — COMMENTED** [review]
Kinda hacky but many things are here :) 

If you can get the rest of the tests to pass, I think we should be good.

File: packages/test-utils/src/wrapper.js (hunk 1 above)
Comment (AtofStryker) [reply packages/test-utils/src/wrapper.js]: @lmiller1990 this would be a comparison by reference if we are considering objects/array. But these should be assigned by reference [here](https://github.com/vuejs/vue-test-utils/blob/dev/packages/test-utils/src/wrapper.js#L709) which makes me think this should be enough to check the reference, unless something is getting mutated under the hood? Might be good to do a value check here to be safe

**Comment by AtofStryker** [comment]
> Kinda hacky but many things are here :)

> If you can get the rest of the tests to pass, I think we should be good.

@lmiller1990 It looks like the [tests](https://app.circleci.com/pipelines/github/vuejs/vue-test-utils/249/workflows/4ee4b0f9-6170-4aa4-ab67-ef6973fcc8b1/jobs/14879) are timing out? I was able to get them to pass locally so I am unsure of the issue

File: test/specs/wrapper/setProps.spec.js [hunk 2]
```diff
@@ -187,6 +187,81 @@ describeWithShallowAndMount('setProps', mountingMethod => {
...
+
+      expect(callback.calledOnce)
+      callback.resetHistory()
+
+      await wrapper.setProps({ propA: 'value' })
+      expect(wrapper.props().propA).to.equal('value')
+      expect(callback.calledOnce)
+    })
+
+    it('invokes watchers with immediate set to "true" with deep objects', async () => {
```
Comment (AtofStryker) [inline test/specs/wrapper/setProps.spec.js:218]: also added a deep equality test here to test the assign by reference in the code to make sure we don't get into some type of continual recursive loop. This looks to work as intended

**Comment by AtofStryker** [comment]
Looks like this [test](https://github.com/vuejs/vue-test-utils/blob/dev/test/specs/wrapper/setProps.spec.js#L77) is hanging?

**Comment by lmiller1990** [comment]
I will take a look at the hanging test. 

I think we should drop support for < Vue 2.5 sometime.

**Comment by dobromir-hristov** [comment]
Heh nice. I would not have thought about this :D

**Comment by AtofStryker** [comment]
@dobromir-hristov I almost wish I didn't think of this 😆 . It's pretty hacky, but wasn't sure of another way, especially since [Watcher](https://github.com/vuejs/vue/blob/dev/src/core/instance/state.js#L357) doesn't seem to do anything with the `immediate` option, except execute the callback 'immediately'

**Comment by AtofStryker** [comment]
@lmiller1990 any luck with that hanging test? Let me know if I can help in any way!

**Comment by lmiller1990** [comment]
I spent a bit of time but I am not sure why is is hanging.

There is some way to skip tests for old Vue versions, you could give that a shot (if that's the reason it's hanging).

**Comment by AtofStryker** [comment]
> I spent a bit of time but I am not sure why is is hanging.
> 
> There is some way to skip tests for old Vue versions, you could give that a shot (if that's the reason it's hanging).

@lmiller1990 is there a config val in CI that handles that? Otherwise, I could just use `xdescribe` on a commit for that particular test and push it up, if that is alright with you

**Comment by lmiller1990** [comment]
@AtofStryker sorry, I mean you can use this helper: https://github.com/vuejs/vue-test-utils/blob/1e3f7fbf7970b0e23715dac6a6cf2a468b0e6e3c/test/specs/wrapper/trigger.spec.js#L112

**Comment by AtofStryker** [comment]
@lmiller1990 I figured out why the test was hanging. I completely forgot the [`$attrs`](https://github.com/vuejs/vue-test-utils/blob/dev/packages/test-utils/src/wrapper.js#L698) use case, which was causing the test to go into an infinite recursion loop, which is bad :grimacing: . I have added an additional check here to compare for `$attrs` if the key doesn't exist on `vm` [here](https://github.com/vuejs/vue-test-utils/pull/1618/commits/fa669fab04c9a36ac2690e79186bb04c426e8ab8#diff-f8b21801dcf5dee76bc9f5f22e227ea8R719). That should make the tests pass and not introduce breaking changes :smile:

File: packages/test-utils/src/wrapper.js [hunk 3]
```diff
@@ -711,7 +711,17 @@ export default class Wrapper implements BaseWrapper {
 
       // $FlowIgnore : Problem with possibly null this.vm
       this.vm.$forceUpdate()
-      return nextTick()
+      return new Promise((resolve, reject) => {
```
Comment (AtofStryker) [inline packages/test-utils/src/wrapper.js]: Since nothing is rejected?
```suggestion
      return new Promise(resolve => {
```

**Comment by lmiller1990** [comment]
Haha, so the test was actually correctly helping us out and catching the `attrs` edge case. Awesome! Great job on this, I will merge and do a release in the next week (probably with your babel dependencies ones that is merged).

**Comment by 99linesofcode** [comment]
@AtofStryker if I may, did you happen to figure out why the next tick caused the `setProps` call to fail?

File: packages/test-utils/src/wrapper.js [hunk 4]
```diff
@@ -711,7 +711,17 @@ export default class Wrapper implements BaseWrapper {
...
-      return nextTick()
+      return new Promise(resolve => {
+        nextTick().then(() => {
+          const isUpdated = Object.keys(data).some(key => {
+            return (
+              // $FlowIgnore : Problem with possibly null this.vm
+              this.vm[key] === data[key] || this.vm.$attrs[key] === data[key]
+            )
+          })
+          return !isUpdated ? this.setProps(data).then(resolve()) : resolve()
```
Comment (TeamDman) [inline packages/test-utils/src/wrapper.js:722]: Shouldn't this be
`.then(() => resolve())`
instead of
`.then(resolve())`
?

File: packages/test-utils/src/wrapper.js [hunk 5]
```diff
@@ -711,7 +711,17 @@ export default class Wrapper implements BaseWrapper {
 
       // $FlowIgnore : Problem with possibly null this.vm
       this.vm.$forceUpdate()
-      return nextTick()
+      return new Promise(resolve => {
+        nextTick().then(() => {
+          const isUpdated = Object.keys(data).some(key => {
```
Comment (TeamDman) [inline packages/test-utils/src/wrapper.js:716]: Shouldn't this use `.every`?
Otherwise, setting new props where one but not all of the new values is the same as the old could falsely signal that the update succeeded.

File: packages/test-utils/src/wrapper.js (hunk 4 above)
Comment (TeamDman) [reply packages/test-utils/src/wrapper.js:722]: **EDIT: ignore this, I forgot that this was overriding the method, so we know it's always returning a promise**
Maybe the typedef could be updated
_____

Additionally, the signature on setProps claims to potentially return void
```typescript
interface BaseWrapper {
	setProps (props: object): Promise<void> | void
}
```

so I'm getting warnings on
```typescript
? this.setProps(data).then(`
```
```
property 'then' does not exist on type 'void | Promise<void>'.
  Property 'then' does not exist on type 'void'
```

maybe I'm missing something, I'm just trying to backport the fix for my project

<details>
<summary>Janky backport function if anyone cares</summary>

```typescript
/**
 * Bug: setProps doesn't play nice when there exists a watcher with immediate:true
 * https://github.com/vuejs/vue-test-utils/pull/1618
 */
function setPropsForSure(
	wrapper: Wrapper<CombinedVueInstance<any, any, any, any, any>>,
	data: any,
) {
	return new Promise(resolve => {
		wrapper.setProps(data);
		wrapper.vm.$nextTick().then(() => {
			const isUpdated = Object.keys(data).every(key => {
				return (
					// $FlowIgnore : Problem with possibly null this.vm
					wrapper.vm[key] === data[key] || wrapper.vm.$attrs[key] === data[key]
				);
			});
			if (isUpdated) {
				return resolve();
			}
			let result = setPropsForSure(wrapper, data);
			if (result) {
				result.then(() => resolve());
			} else {
				resolve();
			}
		});
	});
}
```

</details>

**Comment by vegerot** [comment]
Not sure if this is the culprit, but we have a test that appears to succeed, but never finishes.  i.e the test reports success, but `jest` never exits.  This was not a problem on `1.0.3` but is a problem on every version since.

Not a full repro, but the test is something like

```ts
    it('should use foo\'s dates when bar hasn\'t been set', () => {
      expect.assertions(2);
      wrapper.setProps({
        a: {
          ...a,
          endTime: '2020-01-02',
          startTime: '2020-01-01',
        },
        b: {
          endTime: '2020-01-02',
          startTime: '2020-01-01',
        },
      });
      expect(wrapper.vm.bar).toBeTruthy();

      wrapper.setProps({
        a: {
          ...a,
          endTime: '2020-01-02',
          startTime: '2020-01-01',
        },
        b: {
          endTime: '2020-01-01',
          startTime: '2020-01-01',
        },
      });
      expect(wrapper.vm.bar).toBeFalsy();
    });
```

This test passes very quickly (about one second), but 10 minutes later and jest is still doing...something.

This is fixed by doing

```ts
    it('should use foo\'s dates when bar hasn\'t been set', async () => {
      expect.assertions(2);
      await wrapper.setProps({
        a: {
          ...a,
          endTime: '2020-01-02',
          startTime: '2020-01-01',
        },
        b: {
          endTime: '2020-01-02',
          startTime: '2020-01-01',
        },
      });
      expect(wrapper.vm.bar).toBeTruthy();

      await wrapper.setProps({
        a: {
          ...a,
          endTime: '2020-01-02',
          startTime: '2020-01-01',
        },
        b: {
          endTime: '2020-01-01',
          startTime: '2020-01-01',
        },
      });
      expect(wrapper.vm.bar).toBeFalsy();
    });
```

**Comment by lmiller1990** [comment]
We should update the docs to encourage people to always do `await setProps` - would that solve the problem?

Actually instead of using `setProps` I just write two tests - I find that easier to debug anyway. But we should make sure `setProps` still works well

**Comment by vegerot** [comment]
Well for me, using `await setProps` solved my problem 🤷‍♀️

**Comment by Serhansolo** [comment]
@vegerot, I can't solve the same issue with awaiting. Does anyone know more about this weird issue and why it's happening?
